frontend/frontend/honeylogger/_postgres.py

```diff
@@ -122,21 +122,26 @@ def __str__(self) -> str:
     def __del__(self):
         if self._conn is not None:
             logger.critical(
-                '%s DATA LOST AND DB CONN LEAKED! Logging session was started but never ended.',
+                '%s DATA LOST! Logging session was started but never ended.',
                 self)
+            self._conn_pool.putconn(self._conn)
+            self._conn = None
```

Comment:
This could be dangerous since the pool theoretically could have been closed or something similar once GC runs for this object. However, in our case, the pool is never closed or cleaned up so it should be fine.

Reply:
Yeah, it's pretty much a last resort to avoid dropping connections from the pool (if it were to happen). I think I fixed one race condition in the most recent commit which I think occurred like this:

1. `end()` is called on the logging session since `TransportManager` detects that the SSH is inactive, and then disposes of everything related to it
2. The logging session still has a queued call to some logging method from another thread. Since the `end()` method had been called, the `_conn` is `None` and `_get_connection` gets another connection from the pool
3. `TransportManager` doesn't have a reference to the logging session anymore since it has ended it, but the logging session now has an active connection from the pool that will be lost when the object is garbage collected.

I'm sure there are more race conditions so I think we need to make sure to return the connection if it isn't None when the object is garbage collected. I'll add a check to see if the pool is closed before doing it, that will hopefully make it safe

Reply:
Ohh, that's a very sneaky bug. Yeah, that makes sense and, hopefully, safe and catches any other accidentally open connections.